Accelerating Cloud Native in Telco whitepaper (v1) (#272)
cncf/cnf-wg: doc/whitepaper/Accelerating_Cloud_Native_in_Telco.md

that to attain the envisioned outcome, the entire industry needs to work together to align around key strategic and operational 
principles. Besides building a sound understanding of what it would take for the transformation of CNFs to become cloud native, 
it's also important to emphasize the ecosystem that would support that evolution.

The industry is still maturing and searching for the right formula to reach a cloud native operating model. CNF vendors have not been able to comply with the cloud native and openness requirements of CSPs yet, due to the fact that these requirements are not yet stable and still emerging; however, the reasons for this hesitance can be found in an aversion to giving up a lucrative professional services business model and control over vertical integration. This hesitance cannot override a CSP's need to evolve towards a cloud native architecture, largely based on 12-factors for 
CNFs (see Annex 1 and Reference 5), that can be supported by changing existing commercial models that could greatly benefit CSPs and vendors alike to create a new win-win equilibrium. Vendors must provide open APIs, clear documentation, and cloud native 

> architectures and implementations that empower CSPs with self-service capabilities in the cloud ecosystem. In order for new model to work the vendors and CSPs must provide mutual SLAs: the CSP must guarantee a certain level of quality at the platform layer, while CNF vendors need to guarantee that the application will perform on the platform with SLAs that meet defined KPIs. This will help drive 

Reviewer: nit
```suggestion
architectures and implementations that empower CSPs with self-service capabilities in the cloud ecosystem. In order for the new model to work, vendors and CSPs must provide mutual SLAs: the CSP must guarantee a certain level of quality at the platform layer, while CNF vendors need to guarantee that the application will perform on the platform with SLAs that meet defined KPIs. This will help drive 
```

that move away from closed solutions and professional services.

 As such, we want to highlight major challenges facing cloud native telco transformations today and formulate principles and 
 requirements that will aid the industry in achieving alignment and overcoming obstacles. 

## Challenges in Cloud Native Telco Transformation Today

**Pre-Validation.** Historically, Network Functions have been developed and pre-integrated with well-defined infrastructure, 
which was known in advance. That pre-validation was done by a Network Function vendor and the system was delivered as a 
validated/certified bundle together with performance and stability guarantees. In the cloud native world, there are too many 
permutations which makes it impractical to follow the traditional certification path. However, Cloud Native Network Function (CNF) 
vendors are still sticking to it by picking a small number of opinionated infrastructure flavors (different from vendor to vendor!) 
to pre-validate against, making any infrastructure outside this selection too complicated, too costly, and too slow to deliver for 

> CSPs. This creates problems in the adoption of those CNFs as CSPs generally prefer to each have a unified cloud native infrastructure 

Reviewer 2: I'm not sure about this, Cloud Native infrastructure is about infrastructure implementing the cloud native principle. I think it's important to use the word to emphasize that it's not about cloud (i.e. VM-centric initial world). BTW, it's implicit in the principles you mention. Otherwise, it's perhaps I'm spending most of my time there, and I'm biased. At the end, we are talking about Cloud Native Network function...

> CSPs. This creates problems in the adoption of those CNFs as CSPs generally prefer to each have a unified cloud native infrastructure 

Author: > “Cloud native” is a loaded term. As much as it has been hijacked by marketing departments, it still can be meaningful for engineering and management. To us, it is the evolution of technology in the world where public cloud providers exist.
> Cloud native infrastructure is infrastructure that is hidden behind useful abstractions, controlled by APIs, managed by software, and has the purpose of running applications. Running infrastructure with these traits gives rise to a new pattern for managing that infrastructure in a scalable, efficient way.
> Abstractions are useful when they successfully hide complexity for their consumer. They can enable more complex uses of the technology, but they also limit how the technology is used. They apply to low-level technology, such as how TCP abstracts IP, or higher levels, such as how VMs abstract physical servers. Abstractions should always allow the consumer to “move up the stack” and not reimplement the lower layers.”

(2018) Cloud Native Infrastructure by Justin Garrison & Kris Nova - Chapter 1: What Is Cloud Native Infrastructure?

> CSPs. This creates problems in the adoption of those CNFs as CSPs generally prefer to each have a unified cloud native infrastructure 

Author: <img width="668" alt="image" src="https://github.com/cncf/cnf-wg/assets/26697/3776515d-7749-4b8a-93b6-17139dcf068a">
(2018) Cloud Native Infrastructure by Justin Garrison & Kris Nova

> CSPs. This creates problems in the adoption of those CNFs as CSPs generally prefer to each have a unified cloud native infrastructure 

Author: Maybe we should add this somewhere:

> “Cloud native infrastructure is not a solution for every problem, and it is your respon‐
sibility to know if it is the right solution for your environment (see Chapter 2). How‐
ever, its success is evident in the companies that created the practices and the many
other companies that have adopted the tools that promote these patterns.”
- (2018) Cloud Native Infrastructure by Justin Garrison & Kris Nova - Chapter 1: What Is Cloud Native Infrastructure?

> CSPs. This creates problems in the adoption of those CNFs as CSPs generally prefer to each have a unified cloud native infrastructure 

Author: Cloud-native infrastructure, and specifically infrastructure built using the K8s framework, is designed to promote managing applications in a cloud native way along with ways of running the infrastructure itself in a cloud native manner.  This is important because you are free to use Kubernetes (and other "cloud native" technologies) in a very uncloud-native way.

Our longer-term goal, underlying this whitepaper, is for all layers of the environment to encompass the cloud native principles from infrastructure allocation + management, through the application workloads.

> CSPs. This creates problems in the adoption of those CNFs as CSPs generally prefer to each have a unified cloud native infrastructure 

Author: For the purpose of this whitepaper you can think about cloud native infrastruture as the layer that abstracts the infrastructure at as a service that hyperscalers provide or the similar layer for bare metal private clouds.  

A high-level and generalized list of the layers are: application (including workloads), cloud native infrastructure, infrastructure as a service, infrastructure resources including compute, storage, network. (just as seen in the diagram above)

in the local CSP environment and rather insist on conformance with the pre-validation stack. This practice is unsustainable and 
requires a fresh and flexible local validation approach. Automation (Continuous Testing) is especially important when validating 
frequently released cloud native applications and checking conformance with frequently updated cloud platforms.

**Automation.** In the ongoing pursuit of end-to-end orchestration and deployment, and configuration automation, the Telco 
industry has devised numerous frameworks, models, and standards. Some have achieved considerable success, while others have 
seen varying levels of adoption. However, the cloud native ecosystem, with a focus on GitOps practices, is propelling CNFs 
toward more advanced and automated models.

Many CNFs are still reliant on manual artifact deployment and are rooted in traditional telco methods, such as NETCONF and 
YANG for configuration management. These practices pose significant challenges for CSPs aiming for a fully automated CNF lifecycle. 
Moreover, the ETSI standard follows an imperative top-down approach, often characterized as "fire and forget”. This approach

> doesn't readily support reconciliation and depends on orchestration entities operating externally to Kubernetes "out-of-band." 

Reviewer 2: Good point of clarification as I had also the question on my side. We have to be precise and stating K8S at the very beginning, even if it turns obvious by reading after.

> doesn't readily support reconciliation and depends on orchestration entities operating externally to Kubernetes "out-of-band." 

Author: @pensarguet @vukg can you make a suggest-edit for the introduction statements? 

I think cloud native principles can be applied to other orchestration systems (including VM systems), but the requirements list is focused on K8s so that should be pointed out.

> doesn't readily support reconciliation and depends on orchestration entities operating externally to Kubernetes "out-of-band." 

Reviewer 2: I suggest we could add a final sentence in the intro like this one : In this whitepaper, we are defining Kubernetes as the de-facto runtime environment for hosting the cloud native network functions (CNFs). 
@vukg @taylor @ildikov What's your thought ?

which are not generalized nor easily customizable, as well as with divergent configuration schemas. This all creates further 
complexities in the transition to the declarative and GitOps-driven automation models prevalent in the cloud native ecosystem.

**Dependencies.** Cloud Native applications shall be completely separated from the underlying infrastructure, especially hardware. 
Nevertheless, in practice today there are often very hard dependencies present, be it on specific technologies, or on specific vendor
products. The CNFs often require a specific hardware type or brand (e.g. CPU, NIC) and do not allow for flexibility supported by 
local validation. CNFs are not able to run on any CNCF-certified Kubernetes distributions. Even if the dependency is fulfilled there 
is a lack of attention to those dependencies. For example, a CNF can break because the firmware of the network card was updated, 
which shows that pre-validation of a particular combination of dependencies was not performed and proactive CNF update measures 
have not been taken.  This creates a lot of operational burdens and negatively impacts KPIs for cloud native CNF deployments. 

> **Lifecycle.** Kubernetes occupies a central place in cloud native infrastructure by following the paradigm of ephemeral resources

Reviewer 2: We need to be consistent with what we will end up with based on the previous comment. I prefer to stick to cloud native as end to end is too generic here for me.

> **Lifecycle.** Kubernetes occupies a central place in cloud native infrastructure by following the paradigm of ephemeral resources

Author: @vukg when addressing the term in the pre-validation this sentence should be consistent.

> **Lifecycle.** Kubernetes occupies a central place in cloud native infrastructure by following the paradigm of ephemeral resources

Author: definition added to intro

to capture the traces. In many cases, the CNFs or their microservices run on the same node and their communication does not go
via data center network fabric at all. Furthermore, encryption and mTLS became a de-facto standard for CNFs, so even if tapped, 
network traffic can not be really analyzed and so the purpose of tracing can not be fulfilled. Cloud native tracing mechanisms 
(e.g. eBPF) are unfortunately not helping here as most of the telco-relevant traffic goes via secondary interfaces (Multus) 
which are often directly assigned to the CNF, skipping the host kernel drivers. This is specifically true for user plane CNFs like a UPF, Firewall or Internet Gateway.

**Architecture.** We are witnessing that there are still CNFs that are in their architecture exhibiting properties of Virtualized
Network Functions (VNFs). For example, we see the “pinning” of Pods to specific cluster nodes. 
We also still see 1+1 redundancy models for Pods within the cluster instead of N+1. Although it is technically possible to run such
Network Functions on cloud native infrastructure, this increases the burden of operating them and risks having a negative impact 

> on service quality, as small disruptions which are normal in cloud native infrastructures result in problems within the CNFs. 

Reviewer 2: ok, but the title of the white paper is "Challenges of Cloud Native Telco Transformation today and how to overcome them"

> on service quality, as small disruptions which are normal in cloud native infrastructures result in problems within the CNFs. 

Author: @vukg review this for any updates with the term expansion/changes in the pre-validation

> on service quality, as small disruptions which are normal in cloud native infrastructures result in problems within the CNFs. 

Author: This is now addressed in the introduction: 

>  We also use the term cloud-native infrastructure in broader context for the infrastructure that abstracts Infrastructure-as-a-Service (IaaS) layer, that has Kubernetes in its core with useful API abstractions on top of it as as well as auxiliary systems all as a framework that makes managing applications easier and promotes doing so in a cloud native way. 
> This is important because you are free to use Kubernetes (and other "cloud native" technologies) in a very uncloud-native way. Our longer-term goal, underlying this whitepaper, is for all layers of the environment to encompass the cloud native principles from infrastructure allocation + management, through the application workloads. A more in-depth analysis of these terms can be found in the book "Cloud Native Infrastructure" by Justin Garrison and Chris Nova.

comes to dealing with cluster security-relevant aspects like Roles, ClusterRoles, privileged access, cluster node level access, 
and similar functionalities. We frequently observe that the principle of least privilege is not consistently followed and that 
Roles frequently require rights for everything (“*”) and ClusterRoles are used without real need. CNFs sometimes use problematic 
practices (such as hostPath mounting to write their logs, hostPorts for communication, privilege escalation, running containers 
as root, managing the configuration of the node networking stack, and performing dangerous sysctl calls), none of which are 
allowed in a properly hardened environment. It looks like such CNFs assume that the infrastructure can be consumed from a 
cluster admin perspective without any restrictions, which in realistic circumstances is never the case. Such expectation could
be reasonable in a combo/silo package where CNF and infrastructure come together from one hand as a managed package. However, 
in other cases, CNFs are usually “guests” on the infrastructure and as such must have appropriate security imposed restrictions and 
limitations. 

> **Resilience.** In contrast to traditional expectations within the telecommunications domain, one important property of the cloud 

Author: Architecture can be too broad. Maybe it all needs to be broken down into sections.  I'm resolving this until someone puts a suggestion forward.

infrastructure is that it is imperfect. Cloud infrastructure does not give strict performance and stability guarantees. 
However, it offers mechanisms to applications through which they can achieve a high degree of resilience to this imperfection. 
Yet, repeatedly we experience the cases in which the imperfection of cloud infrastructure has a severe impact on the CNFs to 
the point that complete re-deployment is the only viable solution. Instances of such impact include CNFs completely crashing 
due to the ephemeral storage on the Kubernetes cluster reaching full capacity with logs of that CNF or because write operations 
to persistent volumes could not be performed for a short period of time. This state is unsustainable as such events and situations 
are going to happen in the cloud environment all the time. Therefore, applications including CNFs, which aim to run in the cloud, 
have to account for such events in their design and utilize cloud native mechanisms to maintain robustness consistently or facilitate
automated recovery.

## Principles and Requirements to Enable Progress in Cloud Native Telco Transformation

> These principles and requirements address the systemic challenges listed above. They are intended to serve as the guardrails 

Reviewer 3: The way how this section is structured, it doesn't allow for separation between challenges, principles and requirements. I think it would be beneficial to separate those at one point, to make it clear what the principles are, how they address challenges, and what requirements need to be fulfilled for compliance.

Author: @vukg I like the idea in general, but that seems like a major update.  Maybe that should happen in a v1.1 version of this whitepaper?  @pensarguet @tomkivlin @ashanvbs

Author: @ildikov unless you want to provide a version in a pull request that does that separation?

which every industry/ecosystem player, who aims to engage in cloud native telco transformation together with CSPs, shall respect.
In this way, setting the minimal set of conditions, we want to eliminate main blockers and fragmentation, which is a chronic 
condition in the telco ecosystem, and thus accelerate the path towards benefits for all. We aim to work with existing industry 
initiatives such as Anuket RA2 / GSMA NG.139 Kubernetes Reference Architecture to ensure these principles are developed in 
collaboration with existing communities, and included in existing well-established publications.

1. **Pre-validation.** The pre-validation of CNF needs to be performed against a reference that is common for all players,
   which is the upstream Kubernetes and components from the CNCF ecosystem.
   1. Each CNF shall be validated, including the implementation of any adaptations that may be required, within 4 months of

>    the Kubernetes release.

Reviewer 3: This requirement is a little confusing to me. Does this mean that a CNF has to and can only be validated against Kubernetes release(s) that are no more than 4 months old?

I also wonder if this bullet point might go one step too far into the details for a whitepaper? As something as concrete as '4 months' might get updated in the future, which would create confusions if the whitepaper is not immediately updated.

>    the Kubernetes release.

Reviewer 2: 4 months if we are considering the rolling release of K8S seems to not be a bad timeframe according me. Anyway, whether it is 3 or 4 or 5 months, the level of automation and integration that is needed to perform what is waited should be exactly the same.

   1. Every CNF shall certify adherence to cloud native principles and best practices using CNF Test Suite ([https://github.com/cncf/cnf-testsuite](https://github.com/cncf/cnf-testsuite)) as a vendor-neutral validation tool.
   1. Pre-validating CNF against additional commercial distributions such as OpenShift, Tanzu, Rancher, and Hyperscaler solutions is a plus, but not mandatory.

> 1. **Adaptations.** CNF artifacts (e.g. YAML manifests, Helm charts, NFVO descriptors) shall become open for CSP’s DevOps or

Reviewer 3: Would it be better to say 'made available' as opposed to 'become open'? It reads confusing in the sense whether the expectation is to make these artifacts open source, or just accessible to the teams that will need them.

>    vendor’s delivery team to adapt them to align the deployments to the local specifics of CSP (e.g. Policy, RBAC, Compatibility)

Reviewer 3: ```suggestion
   vendor’s delivery teams to adapt them to align the deployments to the local specifics of CSP (e.g. Policy, RBAC, Compatibility)
```

   without special Change Request or complex processes involving R&D.
    1. Validation of such adapted CNF deployment shall be performed on CSP premises.
    1. Given the successful validation vendor support for such deployment shall be granted.
    1. CNFs shall be modular, microservice-based, open applications and not big “black” boxes.
    1. CNF vendors shall provide all artifacts (Helm charts, CRDs, Operators) passing strict linter in open documentation and
       provide APIs, instead of encapsulating them in proprietary tools.
 
1. **Validation.** CNFs shall be delivered with a series of automated tests that can be used to validate the CNF operation on the spot in
   CSP’s context.

>     1. This validation shall count as only relevant one, preceding any pre-validation or lack of it.

Reviewer 3: This bullet point reads confusing to me. Does this mean that this 'validation' replaces the 'pre-validation', if it was performed at all?

Reviewer 4: It means that if pre-validation that CNF Vendor does in its labs is good indication, but shall not be blocker for CNF release to be deployed in some environment that was not pre-validated. For example CNF vendor has pre-validated on OpenShift that uses K8S 1.26, while CSP has upstream K8S 1.27. The vendor shall not say "I have pre-validated in only on OpenShift with K8S 1.26 so I can not support that CNF on K8S 1.27". Instead, if the validation and e2e tests on CSP premisses are passed, CNF shall be automatically supported in that environment.

    1. The validation shall assure that all artifacts are passing strict linters to prove that portability is assured.
    1. It shall serve as a condition for support and SLA.
    1. The validation shall be a continuous process and shall be instantly done on any change be it on CNF or on the infrastructure
       side.
    1. The validation tests  shall cover CNF basic functionality, lifecycle and disaster recovery

1. **Automation.** CNF deployment and configuration shall be fully automated (“everything as a code”) and done exclusively with

>    declarative cloud native mechanisms like GitOps.

Reviewer 3: I think it would be useful to add references that describe what tools and mechanisms the paper is referring to here, for those who are not yet familiar.

Author: GitOps is not a single product, plugin, or platform. While the practices and patterns in GitOps existed before Cloud Native (and the term GitOps), they happen to be a great match for cloud native applications and infrastructure alike.

Here are some principles for GitOps (as define by the Open GitOps org - https://opengitops.dev/):
- Declarative - A [system](https://github.com/open-gitops/documents/blob/v1.0.0/GLOSSARY.md#software-system) managed by GitOps must have its desired state expressed [declaratively](https://github.com/open-gitops/documents/blob/v1.0.0/GLOSSARY.md#declarative-description).
- Versioned and Immutable - Desired state is [stored](https://github.com/open-gitops/documents/blob/v1.0.0/GLOSSARY.md#state-store) in a way that enforces immutability, versioning and retains a complete version history.
- Pulled Automatically - Software agents automatically pull the desired state declarations from the source.
Continuously Reconciled - Software agents [continuously](https://github.com/open-gitops/documents/blob/v1.0.0/GLOSSARY.md#continuous) observe actual system state and [attempt to apply](https://github.com/open-gitops/documents/blob/v1.0.0/GLOSSARY.md#reconciliation) the desired state.

Here are some GitOps references:
- https://opengitops.dev/
-  https://www.gitops.tech/
- https://github.com/readme/featured/defining-gitops
- https://www.atlassian.com/git/tutorials/gitops
- https://www.redhat.com/architect/understanding-gitops
- https://about.gitlab.com/topics/gitops/

Author: Added as annex 2 at end of doc

>     1. Mainstream open source deployment tools from CNCF ecosystems, like FluxCD or ArgoCD, shall be supported per default.

Reviewer 3: ```suggestion
    1. Mainstream open source deployment tools from the CNCF ecosystem, like FluxCD or ArgoCD, shall be supported per default.
```

    1. Alternatively it shall equally support all the mainstream available x86/amd64 compute hardware with single socket servers as golden standard. 
    1. Local validation, not product policy, shall give an answer if CNF can run as expected on particular hardware or not. 
    1. In case of hard technical dependencies, the vendor of such CNF shall timely pre-validate its CNF against all new releases of
       hardware-related software (e.g. drivers, firmware) and proactively adapt the CNF to avoid the negative impact of dependency in
       production.
    1. Application resource requirements must be configured declaratively. The application must not be statically configured to
       utilize specific resources including devices, nodes, or machines
    1. When performance is a requirement, the application should utilize open standards for adapters where possible at all levels
       they are available. For example support for multiple CNIs vs a single CNI.
    1. To be able to deploy on any CNCF Certified Kubernetes distributions
    1. CNFs make use of standard APIs for infrastructure and platform capabilities.  For example Service Mesh Interface, Ingress,
       CNI, etc.

> 1. **Lifecycle.** CNFs have to be constructed in a way to fully tolerates graceful cluster rolling upgrade procedures without

Reviewer 3: ```suggestion
1. **Lifecycle.** CNFs have to be constructed in a way that fully tolerates graceful cluster rolling upgrade procedures without
```

We, the group of CSPs gathered around Cloud Native Computing Foundation (CNCF)’s Cloud Native Network Function Working Group (CNF WG),
live on the frontlines of this transformation and have gathered valuable experience in this regard. We firmly believe 
that to attain the envisioned outcome, the entire industry needs to work together to align around key strategic and operational 
principles. Besides building a sound understanding of what it would take for the transformation of CNFs to become cloud native, 
it's also important to emphasize the ecosystem that would support that evolution.

The industry is still maturing and searching for the right formula to reach a cloud native operating model. CNF vendors have not been able to comply with the cloud native and openness requirements of CSPs yet, due to the fact that these requirements are not yet stable and still emerging; however, the reasons for this hesitance can be found in an aversion to giving up a lucrative professional services business model and control over vertical integration. This hesitance cannot override a CSP's need to evolve towards a cloud native architecture, largely based on 12-factors for 

> CNFs (see Annex 1 and Reference 5), that can be supported by changing existing commercial models that could greatly benefit CSPs and 
> vendors alike to create a new equilibrium in win-win. Vendors must provide open APIs, clear documentation, and cloud native 

Author: ```suggestion
CNFs (see Annex 1 and Reference 5), that can be supported by changing existing commercial models that could greatly benefit CSPs and vendors alike to create a new win-win equilibrium. Vendors must provide open APIs, clear documentation, and cloud native 
```

Author: @sanatariq610 @pensarguet @vukg @jeffsaelens here is my suggested alternative (in the comments https://github.com/cncf/cnf-wg/pull/272/files#r1392952690)

----->

## Preamble

The document presented here is a product of the initial joint work of several Communication Service Providers (CSPs) who are active in Cloud Native Computing Foundation (CNCF)’s Cloud Native Network Function Working Group (CNF WG), NGMN Alliance, and projects like Linux Foundation (LF) Europe’s Sylva and Linus Foundation Networking (LFN) Anuket. It is a draft that has been published with the goal of inviting feedback from other CSPs and motivating discussion and improvements in the broader telecommunication industry. We are hoping that through public discourse we can make the document more complete, relevant, and ready for final release. If you would like to contribute to the discussion and document please feel free to open an issue or create a pull request.

## Introduction 

The recently published [Cloud Native Manifesto](https://www.ngmn.org/highlight/ngmn-publishes-cloud-native-manifesto.html) 
from Next Generation Mobile Networks (NGMN) Alliance does an excellent job outlining the vision and target picture for 
cloud native telecommunication networks. The transformation towards a cloud native production model has already 

> commenced in many Communication Service Providers (CSPs). Practical challenges and pain points on this journey, 

Author: @vukg @tomkivlin @pensarguet @ashanvbs 👆🏻

> commenced in many Communication Service Providers (CSPs). Practical challenges and pain points on this journey, 

Reviewer 5: @pgoyal01 @taylor that's what this document (the "## Challenges in Cloud Native Telco Transformation Today" section) describes.

which hinder progress towards the target expressed in the NGMN Cloud Native Manifesto, have been identified and 
are being felt. These hindering aspects are especially prominent in the CSPs which are already taking practical 
transformation steps and are trying to closely follow the vision described. 

We, the group of CSPs gathered around Cloud Native Computing Foundation (CNCF)’s Cloud Native Network Function Working Group (CNF WG),
live on the frontlines of this transformation and have gathered valuable experience in this regard. We firmly believe 
that to attain the envisioned outcome, the entire industry needs to work together to align around key strategic and operational 
principles. Besides building a sound understanding of what it would take for the transformation of CNFs to become cloud native, 
it's also important to emphasize the ecosystem that would support that evolution.

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Reviewer 4: An attempt to address this aspect is in requirement section 3, points ii & iii. Can you relate with this?

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Reviewer 6: I agree with the points in those sections, but I don't think they address the concern above. Assuming we are talking the adaptations and validations sections, then I agree with notion of how CNFs are delivered, that the black box approach era is hopefully coming to an end, and that we should validate in customer environments. 

Most of my concern is the lack of concrete day-2 implications. LCM is touched on and well-articulated, but if we use the validation example (pre-validation vs validating on a customer prem), what assurances does the CNF provider (doesn't even have to be a vendor, could be the CDN software a different internal team provides) have that the lab, stage and prod environments have parody? This is not to say that I think the points aren't valid, they 100% are imho. I only think that that the cloud/devOps/SRE teams need to also provide assurances should the devs meet the demands laid out. How many internal teams claim they need to go to a hyperscaler because the NetOps environments don't provide what they need? 

Basicaly, if a dev provides an SLA, what assurances do they get in return should they move away from the blackbox approach?

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Reviewer 4: what about having another mutual acceptance process with clear conditions for validity of SLA?

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Reviewer 6: I like it. It's basically providing an acknowledgement to CNF devs that the environments will meet certain parameters as well. A little give and take.

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Author: @vukg mutual SLA?

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Author: Resolving with updates based on our convo @jeffsaelens .  If you have further concerns please add a new comment with suggestions

> CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

Reviewer 7: While CNF suppliers may not have been able to "comply with the cloud native and openness requirements of CSPs", the reason attributed, with conviction ("reason is"), to such non-compliance may not be valid.

- Has the WG or CSP's verified with the CNF vendors their reason for non-compliance? 
- Until very recently, did CSPs have well-defined and stated requirements? Consistent across the CSPs?

> architecture and solution approach is revenue from professional service that allows them to create business models on CNFs deployment,

Reviewer 6: ```suggestion
systems approach is revenue from professional service that allows them to create business models on CNF deployments,
```

Author: @sanatariq610 @pensarguet @vukg

Reviewer 8: I suggest:  "Solution approach is revenue from professional services that allows them to create business models centered on CNF deployments,"  Refers to the problem as cultural, business, and technology-centric and a model around that has to change so its a new solution approach ...

Reviewer 2: I'm ok with the proposal from Sana

Author: Suggest-edit created for @sanatariq610 https://github.com/cncf/cnf-wg/pull/272/files/14a5a779cb079e27ac5861e5d112a58e73202a08#r1411193399

> architecture and solution approach is revenue from professional service that allows them to create business models on CNFs deployment,

Author: ```suggestion
architecture and solution approach is revenue from professional services that allows them to create business models centered on CNF deployments,,
```
Suggestion from @sanatariq610

> automation, and operationalizing work. There is a need to evolve into cloud native architecture, largely based on 12-factors for 

Reviewer 7: ```suggestion
automation, and operationalizing work. There is a need to evolve network functions into cloud native architecture, largely based on 12-factors for 
```

Reviewer 6: System is likely a more accurate word here.

Author: @jeffsaelens it seems like this phrasing is an area that can specify the need for the entire system to move towards cloud native
- CSP platform + environment + operational practices
- NF provider/integrator applications
- Hyperscaler offerings for telecom orgs

Not only do the NFs need to evolve to work in a cloud native system/architecture, but the system they run on will ned to evolve.  This is a collaboration between all parties to move things forward.

Thoughts on wording here taking that into context (which you were thinking should be addressed in the document)?

Reviewer 8: Fine with suggested modification..

Reviewer 4: +1

Author: This has been reworded in latest update:
> This hesitance cannot override a CSP's need to evolve towards a cloud native architecture, largely based on 12-factors

the intention is to include all layers not just NFs (or networking applications).

> CNFs (see Annex 1 and Reference 5), that can be supported by changing existing commercial models that could greatly benefit CSPs and 

Reviewer 7: "that can" and then "that could" -- this sentence can do with simplification

> vendors alike to create a new equilibrium in win-win. Vendors must provide open APIs, clear documentation, and cloud native 

Reviewer 7: Instead of in context what vendors must provide, it would be beneficial to have a well-defined list in its own section/sub-section.

```suggestion
and vendors alike to create a new win-win. Vendors must provide open APIs, clear documentation, and cloud native 
```

Author: @pgoyal01 I think you added an extra "and"

The "equilibrium" can be used but the phrasing needs updating. Maybe:
> ... that could greatly benefit CSPs and vendors alike to create a new win-win equilibrium.

@pensarguet @vukg

Reviewer 4: I am fine with that.

Reviewer 8: I am good with @pgoyal01 proposal " ... that could greatly benefit CSPs and vendors alike to create a new win-win equilibrium"

Reviewer 2: ok for me as well

Reviewer 4: +1

Reviewer 6: @taylor  looks like there is agreement here. Commit and resolve?

Author: @jeffsaelens the agreement is on the alternative I suggested. I have added a suggest-edit https://github.com/cncf/cnf-wg/pull/272/files#r1411186228

> architectures and implementations that empower CSPs with self-service capabilities in the cloud ecosystem. This will help drive 

Reviewer 7: If "self-service capabilities" are a requirement then the required capabilities need to be specified? 
Is this only for "cloud ecosystem"?

Author: @pgoyal01 I think this is intentionally left open. The vendors can create new self-service capabilities as offerings. The CSPs can request things they want.

Please keep in mind this whitepaper is a follow-up to the NGMN manifesto adding more details and steps, but it is not claiming to have the totality of all requirements from any given CSPs. 

In fact the idea, as far as I know, is that this is a base list which CSPs will each build from based on their own internal requirements.  We can of course expand on the list of items that the majority of CSPs agree are common.

> agility and innovation, and reduce Opex costs within CSP. CNF vendors can monetize the value of openness to evolve business models 

Reviewer 7: How do open APIs help drive "agility" or "reduce OPEX costs"? There are many reasons for high CSP Opex costs including the inability of CSPs to modify their operational processes and other things.

May I suggest, the WP doesn't need to delve into the CNF vendor business models.

Author: @pgoyal01 this is mainly an acknowledgment that vendor's need a business model that can align and work with the CSP's cloud native requirements and path.  A few ideas were put forth, but they are not the focus of the WP of course.  

This whitepaper may not be the place to go into explanations for every claim, as it is primarily CSPs communicating their challenges and what *they want to see* going forward for the transformation. They have asked for open APIs. 

I think we can have an entire whitepaper on why open APIs can help drive "agility" or "reduce OPEX costs". 🤔 💭 

The short response from me on APIs is to start with the sentence they wrote which says APIs will help 
> empower CSPs with self-service capabilities in the cloud ecosystem.

This is just one example and others related items are in the document. self-service is about agility and faster response to changing needs.  Vendor-neutral interoperability to help with the large-scale integration needs of CSPs is another area.  Automation across all of the environments and components. Open APIs help with all of this. Open APIs are part of open standards (often the driver for those standards such as  [rfc9260 Stream Control Transmission Protocol (SCTP)](https://datatracker.ietf.org/doc/html/rfc9260) or the [Kubernetes APIs](https://kubernetes.io/docs/concepts/overview/kubernetes-api/) which being open allows anyone to write extensions and integrate with K8s in a way that ensures the extension+integration should work with every K8s environment adhering to those open standards.

I want to point out that the open standards are driven and continue to stay relevant (and living) because of open  APIs supporting transparent testing and working implementations using those open APIs (platforms + applications).

Reviewer 8: Agree with @pgoyal01

many CNFs have some elements which are rather sensitive to it. CNFs that are not resilient to ephemeral Kubernetes nodes 
(e.g. crashing when cluster scaling or upgrade occurs) lead to service interruptions during the lifecycle operations, which
is not acceptable, such as on SCTP pods due to s1 interface interruptions which leads to unsupported ISSU. It is often the case 
that [Pod Disruption Budgets](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/) are not properly set, the 
consequence of which is either service interruption or lifecycle operations being blocked. 

**Tracing.** The more cloud native transformation progresses, the more challenging it is to perform the e2e protocol tracing 
using traditional mechanisms based on tapping points on the network fabric level. The reason for that is the dynamic nature of 
cloud native workloads including CNFs. When several CNFs run within one large data center the pods could be distributed to any 
of the servers in any of the racks. This means that particular communication can go via multiple network elements and as the 

> traditional tapping setup is not configurable or capable enough,it is practically impossible to create reasonable port mirroring 

Reviewer: nit
```suggestion
traditional tapping setup is not configurable or capable enough, it is practically impossible to create reasonable port mirroring 
```

to capture the traces. In many cases, the CNFs or their microservices run on the same node and their communication does not go
via data center network fabric at all. Furthermore, encryption and mTLS became a de-facto standard for CNFs, so even if tapped, 
network traffic can not be really analyzed and so the purpose of tracing can not be fulfilled. Cloud native tracing mechanisms 
(e.g. eBPF) are unfortunately not helping here as most of the telco-relevant traffic goes via secondary interfaces (Multus) 

> which are not covered by vanilla Kubernetes. 

Reviewer 9: Kubernetes is getting secondary interfaces probably next-year. The problem here is the direct VF device attachment, which skips any management layer.

Author: @nickolaev any suggest edits?

Reviewer 4: that would be great, but realistically it will likely take 2 years or longer until it comes to Beta state usable in production. It is long period over which the requirement could become irrelevant as CNFs could regulate that on its level.

Reviewer 9: ```suggestion
which are often directly assigned to the CNF, skipping the host kernel drivers. This is specifically true for user plane CNFs like a UPF, Firewall or Internet Gateway.
```

Reviewer 4: fine by me!

**Architecture.** We are witnessing that there are still CNFs that are in their architecture exhibiting properties of Virtualized
Network Functions (VNFs). For example, we see the “pinning” of Pods to particular NUMA nodes, or worse to specific cluster nodes. 
We also still see 1+1 redundancy models for Pods within the cluster instead of N+1. Although it is technically possible to run such
Network Functions on cloud native infrastructure, this increases the burden of operating them and risks having a negative impact 
on service quality, as small disruptions which are normal in cloud native infrastructures result in problems within the CNFs. 
Furthermore, the scalability of today’s CNFs is still sub-optimal. In many cases, it still relies on vertical scaling and manual 
interventions. Sudden increases in demand cause performance degradation and even downtime if the system is not dimensioned in 

> advance for that peak load.

Reviewer 6: I think there is some over correction here. Assuming "Pinning" is bad is a miss in my opinion. Physics still gets a vote in how electrons move about the motherboard. Instead of saying something like pinning is bad, I feel focusing on the 1+1 redundancy model mentioned and bad operational practices in general ("statically pinning" workloads for instance) is where the focus should be. Who cares if I pin to a NUMA zone if I can do it declaratively and the app allows for the scheduler to work its magic? App goes down, or the node is tainted, if it can dynamically look for the next 4 physical cores available in a single NUMA boundary and come back up without human intervention, has the app broke the principles?

Reviewer: Or maybe it has to give explanation why it's bad, like `Pinning is bad when your infrastructure doesn't have support for NUMA`

Reviewer 4: I agree in principle with comment from @jeffsaelens. However the borderlines are thin here and is easy to tap into bad operational practices. The goal this is promoting is let the cloud be generic in first line and do not differentiate between the nodes, CPUs, cores etc for functioning. In the domain of performance fine-tuning we could open the room for  some optimization like NUMA etc.

Reviewer 6: Even "cloud friendly" applications have an endless amount of documentation on performance tuning. A poorly deployed database, an unoptimized content delivery platform, etc... are examples of non-NF applications that also expect some level of infra tuning. Simply consuming more resources in the cloud as the answer to all bottlenecks is not a great path forward (or cost effective). One big difference though is that these examples typically can take advantage of what the cloud offers. Tuning doesn't suddenly block you from horizontally scaling, bursting capacity, etc... which is key.

I think we have to walk that fine line between generic infra and garbage infra. There are a lot of "premium" services in the hyperscalers for tweaking performance and ensuring a better outcome. We just don't balk as hard at those because the hyperscalers don't break their consumption model to enable said services. A lot of us have a ton of scar tissue from broken OpenStack and VMware environments that had so many knobs turned that we consistently engineered ourselves into corners. If we don't break the community's principles though, I don't think we should immediately dismiss implementations unless we've tested and documented the best practice against them.

Reviewer 9: @vukg it's probably a miswording, but it's not a problem of "fine-tuning". I'm sure we all understand that a badly placed "data plane" workload (think far from its PCI NIC and/or allocated memory) can dramatically drop its throughput, introduce all kinds of latency and jitter issues and so on. As @jeffsaelens said above, we can not change physics to be cloud-native, but instead, we can expose the physical constraints in a cloud-native, decalrative way.

Reviewer 4: sure, we shall not single out pinning as bad practice per default. As long as it is possible to manage it in declarative way that relies on general capability of underlying cloud-native infra, instead on particular instance, it shall be ok.

have to account for such events in their design and utilize cloud native mechanisms to maintain robustness consistently or facilitate
automated recovery.

## Principles and Requirements to Enable Progress in Cloud Native Telco Transformation

These principles and requirements address the systemic challenges listed above. They are intended to serve as the guardrails 
which every industry/ecosystem player, who aims to engage in cloud native telco transformation together with CSPs, shall respect.
In this way, setting the minimal set of conditions, we want to eliminate main blockers and fragmentation, which is a chronic 
condition in the telco ecosystem, and thus accelerate the path towards benefits for all. We aim to work with existing industry 
initiatives such as Anuket RA2 / GSMA NG.139 Kubernetes Reference Architecture to ensure these principles are developed in 
collaboration with existing communities, and included in existing well-established publications.

1. **Pre-validation.** The pre-validation of CNF needs to be performed against a reference that is common for all players,
   which is the upstream Kubernetes and components from the CNCF ecosystem.

>    1. Each CNF shall be validated, including the implementation of any adaptations that may be required, within 3 months of

Reviewer 4: 3 months might be too ambitious. I suggest to increase to 4 months, the same as K8s release cadence.

>    1. Each CNF shall be validated, including the implementation of any adaptations that may be required, within 3 months of

Author: ```suggestion
   1. Each CNF shall be validated, including the implementation of any adaptations that may be required, within 4 months of
```

@vukg @tomkivlin

Reviewer 5: OK, make if 4.   I am concerned though, as I suspect many CNFs already do this.  It is the next phase of validation against e.g. Tanzu, OpenShift, etc. etc. etc. that takes the time that needs reducing.

   1. Every CNF shall certify adherence to cloud native principles and best practices using CNF Test Suite ([https://github.com/cncf/cnf-testsuite](https://github.com/cncf/cnf-testsuite)) as a vendor-neutral validation tool.
   1. Pre-validating CNF against additional commercial distributions such as OpenShift, Tanzu, Rancher, and Hyperscaler solutions is a plus, but not mandatory.

1. **Adaptations.** CNF artifacts (e.g. YAML manifests, Helm charts, NFVO descriptors) shall become open for CSP’s DevOps or
   vendor’s delivery team to adapt them to align the deployments to the local specifics of CSP (e.g. Policy, RBAC, Compatibility)
   without special Change Request or complex processes involving R&D.
    1. Validation of such adapted CNF deployment shall be performed on CSP premises.
    1. Given the successful validation vendor support for such deployment shall be granted.
    1. CNFs shall be modular, microservice-based, open applications and not big “black” boxes.
    1. CNF vendors shall provide all artifacts (Helm charts, CRDs, Operators) passing strict linter in open documentation and
       provide APIs, instead of encapsulating them in proprietary tools.
 

> 1. **Validation.** CNFs shall be delivered with a series of automated tests that can be used to validate the CNF on the spot in

Reviewer 9: ```suggestion
1. **Validation.** CNFs shall be delivered with a series of automated tests that can be used to validate the CNF operation on the spot in
```

   CSP’s context.
    1. This validation shall count as only relevant one, preceding any pre-validation or lack of it.
    1. The validation shall assure that all artifacts are passing strict linters to prove that portability is assured.
    1. It shall serve as a condition for support and SLA.
    1. The validation shall be a continuous process and shall be instantly done on any change be it on CNF or on the infrastructure

>        side.

Reviewer 9: ```suggestion
       side.
    1. The validation tests  shall cover CNF basic functionality, lifecycle and disaster recovery
```

Author: @nickolaev I like the idea of this one.. it might be too vague though.  Could it be expanded and each point be covered?

1. **Automation.** CNF deployment and configuration shall be fully automated (“everything as a code”) and done exclusively with
   declarative cloud native mechanisms like GitOps.
    1. Mainstream open source deployment tools from CNCF ecosystems, like FluxCD or ArgoCD, shall be supported per default.

>     1. All configurations shall be done via Configmaps and/or similar cloud native constructs (eg. Kubernetes Resource Models) 

Reviewer: nit:
```suggestion
    1. All configurations shall be done via Configmaps and/or similar cloud native constructs (e.g. Kubernetes Resource Models) 
```

    1. CNF is allowed to use traditional telco mechanisms internally as a transition step, however, that should be fully
       encapsulated and abstracted away.

>     1. Microservices should be loosely coupled with NO tight dependency on each other, to ensure scalability and ease of deployment,

Reviewer: A little redundant
```suggestion
    1. Microservices should be loosely coupled (with NO tight dependency on each other) to ensure scalability and ease of deployment,
```

       e.g. without the need to wait for NETCONF day-1 configuration till further microservices get deployed.
    1. Artifacts are delivered via OCI-compliant repositories.
    1. The CNF LCM should be described declaratively and support continuous intent-based deployments for example IP address
       assignment during deployment. 
    1. New released software version (CNF/microservices) includes machine-readable code to run health checks.
    1. Release notes and impact reports should be included as machine-readable code in every published release.
    1. CSP-internal automation pipeline shall be allowed to hook into the vendor software delivery solution (e.g. to subscribe to
       CNF releases).
    1. Artifacts delivered with CNFs (e.g. Helm charts) shall be customizable for efficient multi-purpose deployments.
    1. CNF configuration schemas have to follow the standards that shall be aligned among CSPs and vendors.

1. **Dependencies.** The CNFs shall be completely independent from underlying infrastructure platforms.

>     1. Alternatively it shall equally support all the mainstream available hardware. 

Reviewer 9: To me, this is too broad term `mainstream available hardware`? 

- I am sure in a telco DC, one can find hardware that is not typically available in any "mainstream DC". Think of a Smart NIC or an FPGA card. 
- If we get down to the CPU - are we expecting every CNF to be equally performant on an AArch64 server?

Author: @nickolaev have suggestions to find a compromise in wording? Something about being vendor neutral between options for the telco DC hardware needed? 

@vukg @pensarguet @tomkivlin

Reviewer 9: I can propose to have a separate paragraph describing "Mainstream Telco Hardware". Let me share some thoughts:

First I think we should explicitly call x86-64/amd64 as a defacto standard in the DC server platforms. Any CPU from 2019 and newer should be supported.
Then we can pin it to a 2-socket, NUMA-enabled HW architecture.
Regarding NICs - to me, the golden standard here is the DPDK drivers list (https://doc.dpdk.org/guides/nics/). Any decent CNF should be able to run with any NIC that is based on these chips. Of course, those that are x86-64 compatible. Even if the CNF can take advantage of a chip-specific HW acceleration, it should be able to deliver at least 60% of its fine-tuned performance on any of the said NICs.
Regarding more specific HW accelerators - these should be an optional requirement. The CNF should provide a decent software implementation to replace them, which can deliver at least 25% of the HW accelerated performance.

Reviewer 5: > I can propose to have a separate paragraph describing "Mainstream Telco Hardware". Let me share some thoughts:
> 
> First I think we should explicitly call x86-64/amd64 as a defacto standard in the DC server platforms. Any CPU from 2019 and newer should be supported. Then we can pin it to a 2-socket, NUMA-enabled HW architecture. Regarding NICs - to me, the golden standard here is the DPDK drivers list (https://doc.dpdk.org/guides/nics/). Any decent CNF should be able to run with any NIC that is based on these chips. Of course, those that are x86-64 compatible. Even if the CNF can take advantage of a chip-specific HW acceleration, it should be able to deliver at least 60% of its fine-tuned performance on any of the said NICs. Regarding more specific HW accelerators - these should be an optional requirement. The CNF should provide a decent software implementation to replace them, which can deliver at least 25% of the HW accelerated performance.

Agree re. the NICs. Disagree re. the 2-socket.  There are many benefits to single socket servers, in particular around not having to mess around with NUMA pinning etc.

Reviewer: What about to use `Commercial Off-The-Shelf (COTS)` term instead of `Mainstream`?

Reviewer 4: +1 for NIC, but for CPU I would set single socket x86/amd64 as golden standard

Author: @electrocucaracha @nickolaev please check latest updates from @vukg.  If that does not resolve these items please make a suggest edit.

    1. Local validation, not product policy, shall give an answer if CNF can run as expected on particular hardware or not. 
    1. In case of hard technical dependencies, the vendor of such CNF shall timely pre-validate its CNF against all new releases of
       hardware-related software (e.g. drivers, firmware) and proactively adapt the CNF to avoid the negative impact of dependency in

>        production.

Reviewer 9: This sounds good in theory, but I am not sure how it will work in practice. It is very often when different NIC vendors (or OEMs) are providing their firmware releases while using the same packet processing chip. So the only option that the CNF vendor has is to pin a particular SKU in their requirements.

Reviewer 4: Do you refer here to HW SKU or SW SKU or both?

Reviewer 9: I was thinking HW when I wrote it, but maybe SW makes sense too.

Author: @vukg @nickolaev did latest update cover these?

    1. Application resource requirements must be configured declaratively. The application must not be statically configured to
       utilize specific resources including devices, nodes, or machines

>     1. When performance is a requirement, the application should utilize open standards for adapters where possible at all levels

Reviewer: I'd add a request for defining `limits` and `request` values on their Pod definitions

Author: @electrocucaracha suggest edit please.

>        they are available. For example support for multiple CNIs vs a single CNI. Support for SR-IOV and…

Author: ```suggestion
       they are available. For example support for multiple CNIs vs a single CNI.
```

other example not complete. removed

Author: @ashanvbs https://github.com/cncf/cnf-wg/pull/272/files#r1382126142

    1. To be able to deploy on any CNCF Certified Kubernetes distributions
    1. CNFs make use of standard APIs for infrastructure and platform capabilities.  For example Service Mesh Interface, Ingress,
       CNI, etc.

1. **Lifecycle.** CNFs have to be constructed in a way to fully tolerates graceful cluster rolling upgrade procedures without
   blocking them and without service interruptions.
    1. CNF shall implement N+1 redundancy mechanisms.
    1. CNF shall rely on mechanisms around PodDisruptionBudgets to secure the conditions for itself to run uninterrupted during
       lifecycle procedures.
    1. The application must function properly if it is rescheduled to other nodes without interruption of its services (e.g. move
       user sessions without interruptions).

>     1. Upgrade procedures on CNF shall also follow rolling-upgrade principles and shall be done in-service

Author: ```suggestion
    1. Upgrade procedures on CNF shall also follow rolling-upgrade principles and shall be done in-service.
```

    1. Microservices should be small and independently deployable units of functionality. 
    1. Any common functionality (e.g. observability, access management) should be provided as a centralized service that can be
       reused. CNFs should share databases, load balancers, business logic, and common services and become fully disaggregated. 
    1. A microservice should have a single, well-defined responsibility, and should only communicate with other microservices to
       accomplish tasks.
    1. The data-processing logic should be in the microservices themselves, rather than in a centralized hub, to ensure scalability
       and reliability.
    1. Microservices should be designed to dynamically scale up or down in response to changes in demand, to ensure that the
       application remains responsive and available using "Horizontal Pod Autoscaling" (HPA).
    1. Each microservice should log information and expose metrics about its performance and usage, which can be used to identify
       and diagnose issues.
    1. CNFs should expose their state (e.g. health) in a cloud native way.

>     1.  CNFs can share databases, load balancers, business logic, and common services and become fully disaggregated.

Reviewer: nit:
```suggestion
    1. CNFs can share databases, load balancers, business logic, and common services and become fully disaggregated.
```

    1. The application must be isolated with Namespaces and not use the default namespace.

1. **Resilience.**  High-quality CNFs should be resilient to underlying infrastructure issues including complete failure,
   meaning that instead of completely failing, they note that something is wrong, log the errors, etc, and then return to
   full working order upon restoration of the underlying infrastructure service/resource.
    1. CNFs use cloud native principles (e.g. Kubernetes capabilities) when implementing resilience architecture without.
       fully focusing on 3GPP (e.g. CNFs do not rely solely on robust infrastructure).
    1. CNFs are resilient to network corruption / poor quality network connection (e.g. packet drops, high latency, etc.).
    1. CNFs are resilient to complete loss of network connection.
    1. CNFs are resilient to poor-quality storage connection (e.g. high latency, read/write performance degradation, etc.).
    1. CNFs are resilient to complete loss of storage connection.
    1. CNFs are resilient to storage disks being full.
    1. CNFs are resilient to CPU stresses.
    1. CNFs are resilient to Memory stresses.

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Author: @nickolaev at the moment you could think of this as the entire list... a test suite covering all items could break things into levels of importance/priority. A certification using the testing could have levels as well.

Maybe the wording needs to use "MUST, SHOULD, MAY".  @tomkivlin @Nery44 @pensarguet @vukg @electrocucaracha

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Reviewer 4: it is actually referring more to ability of CNF to recover on its own after problematic condition occur instead of remaining broken.

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Reviewer 6: I think this ties into the earlier comments around SLAs, Primes and the like. If the infra is providing a certain level of guarantee, then the CNFs should as well from a fault tolerance standpoint. 

Example, we talk about K8s being the defacto here, but we are sort of assuming operators have deployed it in a sane way. K8s CP is expected to handle a node failure, but that is on the pretense that someone has provided the nodes needed for a taint to take place and service migrated. If the infra is in fact meeting "infra level SLAs" though then it's also reasonable to expect for the CNF to have a solid level of fault tolerance. To support @vukg 's point, if a node fails, the CNF should be able to handle the pods being rescheduled and allow for services to be reestablished.

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Reviewer 5: > @nickolaev at the moment you could think of this as the entire list... a test suite covering all items could break things into levels of importance/priority. A certification using the testing could have levels as well.
> 
> Maybe the wording needs to use "MUST, SHOULD, MAY". @tomkivlin @Nery44 @pensarguet @vukg @electrocucaracha

I would read this list as an example, as you say even a "perfect" list.  As I added at the top of this section ([here](https://github.com/cncf/cnf-wg/blob/whitepaper-accelerating-cloud-native-in-telco/doc/whitepaper/Accelerating_Cloud_Native_in_Telco.md?plain=1#L157-L159)), we need to work with established communities and publications to get them discussed and added, rather than starting Yet Another List.

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Reviewer 4: looks good!

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Reviewer 2: I do agree with @vukg

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Author: @nickolaev @vukg @jeffsaelens The latest updates covered some mutual accountability for the transformation.. which includes SLA stuff.. do we need any more changes here? If so can someone please provide a suggest edit.  Otherwise I want to resolve this.

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Reviewer 6: @vukg 's recent additions meet what I was after. Other than changing "are" to "should be" or "must be" I'm fine with things as they are. This is ultimately a list of what "ideal" looks like to a CSP, so I'm not sure watering it down gains us anything. 

The notion of the platform also providing guarantees I think makes it reasonable to expect things from the software in return.

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).

Author: we can probably change a lot of the wording to be must, should, optional etc. let's do that in another round

  enabling continuous deployment for maximum agility. Telcos can deliver features and upgrades faster by implementing
  DevOps practices over a CI/CD pipeline.
* Open: Cloud native applications need to be orchestrated, run as a service, and expose themselves via RESTful interfaces.
  Nephio will achieve this goal by enabling third-party automation tools to reconfigure the application and achieve any
  type of orchestration use case. 
* Traceable: Cloud native applications need to support real-time troubleshooting through Open APIs with telemetry-compatible
  tracing. 
* Loggable: Cloud native applications need to support uniform logging for consistency and access to network-wide logs.

## ACKNOWLEDGEMENTS

Special acknowledgments go to the following Communication Service Providers, who have contributed to this whitepaper: 

* Bell Canada 
    * Daniel Bernier, Technical Director
    * Roger Lupien, Sr. Mgr - Enterprise Architecture, Cloud Transformation
* Charter Communications
    * Mohammad Zebetian, Head of Cloud, Network, Edge, and Infrastructure Architecture

> * Deutsche Telekom Technik

Reviewer 4: Pls correct this section and replace it with:

Deutsche Telekom (DT)
    * Vuk Gojnic, Cloud Native Telco Platform Lead, DT Technik
    * Nathan Rader, VP Service and Capability Exposure, DTAG

> * Deutsche Telekom Technik

Author: ```suggestion
* Deutsche Telekom (DT)
```

Author: @vukg ☝🏻 👇🏻

>     * Vuk Gojnic, Cloud Native Telco Platform Lead
>     * Nathan Rader, VP Service and Capability Exposure

Author: ```suggestion
    * Vuk Gojnic, Cloud Native Telco Platform Lead, DT Technik
    * Nathan Rader, VP Service and Capability Exposure, DTAG
```

commenced in many Communication Service Providers (CSPs). Practical challenges and pain points on this journey, 
which hinder progress towards the target expressed in the NGMN Cloud Native Manifesto, have been identified and 
are being felt. These hindering aspects are especially prominent in the CSPs which are already taking practical 
transformation steps and are trying to closely follow the vision described. 

We, the group of CSPs gathered around Cloud Native Computing Foundation (CNCF)’s Cloud Native Network Function Working Group (CNF WG),
live on the frontlines of this transformation and have gathered valuable experience in this regard. We firmly believe 
that to attain the envisioned outcome, the entire industry needs to work together to align around key strategic and operational 
principles. Besides building a sound understanding of what it would take for the transformation of CNFs to become cloud native, 
it's also important to emphasize the ecosystem that would support that evolution.

CNF suppliers have not been able to comply with the cloud native and openness requirements of CSPs and one reason for that closed 

> architecture and solution approach is revenue from professional service that allows them to create business models on CNFs deployment,

Author: See newer version of whitepaper

have to account for such events in their design and utilize cloud native mechanisms to maintain robustness consistently or facilitate
automated recovery.

## Principles and Requirements to Enable Progress in Cloud Native Telco Transformation

These principles and requirements address the systemic challenges listed above. They are intended to serve as the guardrails 
which every industry/ecosystem player, who aims to engage in cloud native telco transformation together with CSPs, shall respect.
In this way, setting the minimal set of conditions, we want to eliminate main blockers and fragmentation, which is a chronic 
condition in the telco ecosystem, and thus accelerate the path towards benefits for all. We aim to work with existing industry 
initiatives such as Anuket RA2 / GSMA NG.139 Kubernetes Reference Architecture to ensure these principles are developed in 
collaboration with existing communities, and included in existing well-established publications.

1. **Pre-validation.** The pre-validation of CNF needs to be performed against a reference that is common for all players,
   which is the upstream Kubernetes and components from the CNCF ecosystem.

>    1. Each CNF shall be validated, including the implementation of any adaptations that may be required, within 3 months of

Author: @SaadUSheikh See comments from @pensarguet and @vukg. Also read newest version of paper

   vendor’s delivery team to adapt them to align the deployments to the local specifics of CSP (e.g. Policy, RBAC, Compatibility)
   without special Change Request or complex processes involving R&D.
    1. Validation of such adapted CNF deployment shall be performed on CSP premises.
    1. Given the successful validation vendor support for such deployment shall be granted.
    1. CNFs shall be modular, microservice-based, open applications and not big “black” boxes.
    1. CNF vendors shall provide all artifacts (Helm charts, CRDs, Operators) passing strict linter in open documentation and
       provide APIs, instead of encapsulating them in proprietary tools.
 
1. **Validation.** CNFs shall be delivered with a series of automated tests that can be used to validate the CNF on the spot in
   CSP’s context.
    1. This validation shall count as only relevant one, preceding any pre-validation or lack of it.
    1. The validation shall assure that all artifacts are passing strict linters to prove that portability is assured.
    1. It shall serve as a condition for support and SLA.

>     1. The validation shall be a continuous process and shall be instantly done on any change be it on CNF or on the infrastructure

Author: If you have suggestions please specify those clearly.

       CNF releases).
    1. Artifacts delivered with CNFs (e.g. Helm charts) shall be customizable for efficient multi-purpose deployments.
    1. CNF configuration schemas have to follow the standards that shall be aligned among CSPs and vendors.

1. **Dependencies.** The CNFs shall be completely independent from underlying infrastructure platforms.
    1. Alternatively it shall equally support all the mainstream available hardware. 
    1. Local validation, not product policy, shall give an answer if CNF can run as expected on particular hardware or not. 
    1. In case of hard technical dependencies, the vendor of such CNF shall timely pre-validate its CNF against all new releases of
       hardware-related software (e.g. drivers, firmware) and proactively adapt the CNF to avoid the negative impact of dependency in
       production.
    1. Application resource requirements must be configured declaratively. The application must not be statically configured to
       utilize specific resources including devices, nodes, or machines
    1. When performance is a requirement, the application should utilize open standards for adapters where possible at all levels

>        they are available. For example support for multiple CNIs vs a single CNI. Support for SR-IOV and…

Reviewer 10: Remove "Support for SR-IOV and…"

Author: @ashanvbs you can use the suggest edit capability 
<img width="394" alt="image" src="https://github.com/cncf/cnf-wg/assets/26697/cbcd37fe-aa90-4486-a662-165ebabe5646">

Author: Also are we just missing text and need to add the rest @ashanvbs?

Reviewer 10: No its already covered

Author: So resolve this?

    1. To be able to deploy on any CNCF Certified Kubernetes distributions
    1. CNFs make use of standard APIs for infrastructure and platform capabilities.  For example Service Mesh Interface, Ingress,
       CNI, etc.

1. **Lifecycle.** CNFs have to be constructed in a way to fully tolerates graceful cluster rolling upgrade procedures without
   blocking them and without service interruptions.
    1. CNF shall implement N+1 redundancy mechanisms.
    1. CNF shall rely on mechanisms around PodDisruptionBudgets to secure the conditions for itself to run uninterrupted during
       lifecycle procedures.
    1. The application must function properly if it is rescheduled to other nodes without interruption of its services (e.g. move

>        user sessions without interruptions)

Reviewer 10: add full stop

>        user sessions without interruptions)

Author: ```suggestion
       user sessions without interruptions).
```

>     1. Upgrade procedures on CNF shall also follow rolling-upgrade principles and shall be done in-service

Reviewer 10: add full stop

   Reference 3) and in a way not depend on any particular cluster node or reasonably small group of cluster nodes.
    1. Microservices should be small and independently deployable units of functionality. 
    1. Any common functionality (e.g. observability, access management) should be provided as a centralized service that can be
       reused. CNFs should share databases, load balancers, business logic, and common services and become fully disaggregated. 
    1. A microservice should have a single, well-defined responsibility, and should only communicate with other microservices to
       accomplish tasks.
    1. The data-processing logic should be in the microservices themselves, rather than in a centralized hub, to ensure scalability
       and reliability.
    1. Microservices should be designed to dynamically scale up or down in response to changes in demand, to ensure that the
       application remains responsive and available using "Horizontal Pod Autoscaling" (HPA).
    1. Each microservice should log information and expose metrics about its performance and usage, which can be used to identify
       and diagnose issues.

>     1. CNFs should expose their state (e.g. health) in a cloud native way

Reviewer 10: add full stop

>     1. CNFs should expose their state (e.g. health) in a cloud native way

Author: ```suggestion
    1. CNFs should expose their state (e.g. health) in a cloud native way.
```

    1.  CNFs can share databases, load balancers, business logic, and common services and become fully disaggregated.
    1. CNF has to tolerate automatic scaling at the node and container level by the Kubernetes orchestrator.
    1. CNF has to support self-healing.

1. **Security.** In order to run in a generic cloud native environment, CNFs have to strip down their expectations and require
   exactly the minimum rights that are needed for functioning. 
    1. Any practice that poses the risk such as usage of hostPaths, privilege escalations, root containers, etc. needs either to
       be eliminated or replaced with an alternative cloud native approach.
    1. The application must adhere to cluster policies enforced by the cluster manager including overriding its default policies

>     1. _The application should follow the Principle of Least Privilege_

Reviewer 10: remove _italic_ format

>     1. _The application should follow the Principle of Least Privilege_

Author: ```suggestion
    1. The application should follow the Principle of Least Privilege
```
Removing italic at @ashanvbs suggestion

>     1. RBAC definitions must declare the minimal set of rights needed for the application to function The application should not
>        request open-ended / all rights in its RBAC definitions
>     1. Applications should not require privileges to run including privileged pods and cluster roles
>     1. Applications that require privileges must declare which components require privileges in both machine-readable and
>        human-readable open formats

Author: ```suggestion
    1. RBAC definitions must declare the minimal set of rights needed for the application to function The application should not request open-ended / all rights in its RBAC definitions.
    1. Applications should not require privileges to run including privileged pods and cluster roles.
    1. Applications that require privileges must declare which components require privileges in both machine-readable and human-readable open formats.
```

full-stop (period)

>     1. The application must be isolated with Namespaces and not use the default namespace

Reviewer 10: add full stop

>     1. The application must be isolated with Namespaces and not use the default namespace

Author: ```suggestion
    1. The application must be isolated with Namespaces and not use the default namespace.
```

1. **Resilience.**  High-quality CNFs should be resilient to underlying infrastructure issues including complete failure,
   meaning that instead of completely failing, they note that something is wrong, log the errors, etc, and then return to
   full working order upon restoration of the underlying infrastructure service/resource.

>     1. CNFs use cloud native principles (e.g. Kubernetes capabilities) when implementing resilience architecture without

Reviewer 10: add full stop

>     1. CNFs use cloud native principles (e.g. Kubernetes capabilities) when implementing resilience architecture without

Author: ```suggestion
    1. CNFs use cloud native principles (e.g. Kubernetes capabilities) when implementing resilience architecture without.
```

       fully focusing on 3GPP (e.g. CNFs do not rely solely on robust infrastructure)
    1. CNFs are resilient to network corruption / poor quality network connection (e.g. packet drops, high latency, etc.).
    1. CNFs are resilient to complete loss of network connection.
    1. CNFs are resilient to poor-quality storage connection (e.g. high latency, read/write performance degradation, etc.)
    1. CNFs are resilient to complete loss of storage connection.
    1. CNFs are resilient to storage disks being full
    1. CNFs are resilient to CPU stresses
    1. CNFs are resilient to Memory stresses

>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure)

Reviewer 10: add full stop

>        fully focusing on 3GPP (e.g. CNFs do not rely solely on robust infrastructure)
>     1. CNFs are resilient to network corruption / poor quality network connection (e.g. packet drops, high latency, etc.).
>     1. CNFs are resilient to complete loss of network connection.
>     1. CNFs are resilient to poor-quality storage connection (e.g. high latency, read/write performance degradation, etc.)
>     1. CNFs are resilient to complete loss of storage connection.
>     1. CNFs are resilient to storage disks being full
>     1. CNFs are resilient to CPU stresses
>     1. CNFs are resilient to Memory stresses
>     1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure)

Author: ```suggestion
       fully focusing on 3GPP (e.g. CNFs do not rely solely on robust infrastructure).
    1. CNFs are resilient to network corruption / poor quality network connection (e.g. packet drops, high latency, etc.).
    1. CNFs are resilient to complete loss of network connection.
    1. CNFs are resilient to poor-quality storage connection (e.g. high latency, read/write performance degradation, etc.).
    1. CNFs are resilient to complete loss of storage connection.
    1. CNFs are resilient to storage disks being full.
    1. CNFs are resilient to CPU stresses.
    1. CNFs are resilient to Memory stresses.
    1. CNFs are resilient to the complete loss of underlying infrastructure resources (e.g. a node failure).
```

   1. Every CNF shall certify adherence to cloud native principles and best practices using CNF Test Suite ([https://github.com/cncf/cnf-testsuite](https://github.com/cncf/cnf-testsuite)) as a vendor-neutral validation tool.
   1. Pre-validating CNF against additional commercial distributions such as OpenShift, Tanzu, Rancher, and Hyperscaler solutions is a plus, but not mandatory.

> 1. **Adaptations.** It shall possible for CSP’s DevOps or vendor’s delivery team to adapt CNF artifacts (e.g. YAML manifests, Helm charts, NFVO descriptors) to align the deployments to the local specifics of CSP (e.g. Policy, RBAC, Compatibility) without special Change Request or complex processes involving R&D.

Reviewer 11: ```suggestion
1. **Adaptations.** It shall be possible for CSP’s DevOps or vendor’s delivery teams to adapt CNF artifacts (e.g. YAML manifests, Helm charts, NFVO descriptors) to align the deployments to the local specifics of CSP (e.g. Policy, RBAC, Compatibility) without special Change Requests or involving complex R&D processes.
```
